cmd/clusterctl/client/yamlprocessor/simple_processor.go

```diff
@@ -51,21 +51,75 @@ func (tp *SimpleProcessor) GetTemplateName(_, flavor string) string {
 	return name
 }
 
-// GetVariables returns a list of the variables specified in the yaml.
-func (tp *SimpleProcessor) GetVariables(rawArtifact []byte) ([]string, error) {
+// Process processes the template blob of bytes and will return the final
+// yaml with values retrieved from the values getter.
+func (tp *SimpleProcessor) GetVariables(rawArtifact []byte) ([]VariableMetadata, error) {
 	strArtifact := convertLegacyVars(string(rawArtifact))
+	variables, err := getDefaultValues(strArtifact)
+	if err != nil {
+		return nil, err
+	}
+	res := make([]VariableMetadata, 0)
+	for _, v := range variables {
+		if v != nil {
+			res = append(res, v)
+		}
+	}
+	sort.Sort(ByName(res))
+	return res, nil
+}
 
-	variables, err := inspectVariables(strArtifact)
+func getDefaultValues(data string) (map[string]VariableMetadata, error) {
+	t, err := parse.Parse(data)
 	if err != nil {
 		return nil, err
 	}
+	seenMap := map[string]VariableMetadata{}
+	getAllDefault(t.Root, seenMap)
+	return seenMap, nil
+}
 
-	varNames := make([]string, 0, len(variables))
-	for k := range variables {
-		varNames = append(varNames, k)
+func getAllDefault(root parse.Node, seenMap map[string]VariableMetadata) {
+	st := ""
+	switch v := root.(type) {
+	case *parse.ListNode:
+		// iterate through the list node.
+		for _, ln := range v.Nodes {
+			getAllDefault(ln, seenMap)
+		}
+	case *parse.FuncNode:
+		if _, ok := seenMap[v.Param]; !ok {
+			if len(v.Args) > 0 {
+				for _, l := range v.Args {
+					st += getString(l)
```

Comment:
Apologies for the delayed review. Just a quick clarification as I was looking into the parse library tests, it seems like it does various kinds of [string operations](https://github.com/drone/envsubst/blob/490366e43a3ce3dafd1b9584bc7be9a3f5e899b2/parse/parse_test.go#L159).  Should we check for `name == ":"` here to get the default values?

Reply:
I'm not an expert of the envsubst, could you provide some concrete examples of what this means in a template?

Reply:
example:
```
±  |master → origin ↑331 {2} ?:1 ✗| → echo test: ${FOO:=foo} | envsubst
test: foo

 2021-04-14 21:19:16 ⌚ |ruby-2.6.3| sankarans-a01 in ~/workspace/cluster-api-provider-azure
±  |master → origin ↑331 {2} ?:1 ✗| → export FOO=foo && echo test: ${FOO/foo/bar} | envsubst
test: bar

 2021-04-14 21:19:22 ⌚ |ruby-2.6.3| sankarans-a01 in ~/workspace/cluster-api-provider-azure
±  |master → origin ↑331 {2} ?:1 ✗| → export FOO= && echo test: ${FOO/foo/bar} | envsubst
test:
```
IIUC, I believe for case 2 and 3, `len(v.Args) > 0` will be true, and `getString` will return `foobar`. This will mean the default value for `FOO` is `foobar` whereas there is no default value in this case. This will be true for other string operations that envsubst supports. Does that make sense?

Reply:
@shysank thanks for the explanation.
Yes, this function should detect only default operations in the same way envsubst does.

Now if I got it right from 
https://github.com/drone/envsubst/blob/490366e43a3ce3dafd1b9584bc7be9a3f5e899b2/template.go#L150-L153

All the following operators ("=", ":=", ":-" ":?", ":+", "-", "+") acts as a default operators and the default value is always `strings.Join(args, "")`

https://github.com/drone/envsubst/blob/490366e43a3ce3dafd1b9584bc7be9a3f5e899b2/funcs.go#L57

Reply:
Yes, I think you're right. We should check for `v.Name` to be in any one of the above operators. Wondering why there are so many operators for defaulting though 🤔